File: GuideObjC/GuideObjC/ViewControllers/UsersController/UsersController.m
//
//  UsersController.m
//  GuideObjC
//
//  Created by LuzanovRoman on 21.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "UsersController.h"
#import "UserProfileController.h"
#import "UIColor+Additions.h"
#import "SearchDownloader.h"
#import "EdgeDownloader.h"
#import "ProgressCell.h"
#import "UserCell.h"
#import "Follows.h"
#import "EGF2.h"

@interface UsersController ()
@property (retain, nonatomic) IBOutlet UIBarButtonItem *searchButton;
@property (retain, nonatomic) BaseDownloader *activeDownloader;
@property (retain, nonatomic) SearchDownloader *searching;
@property (retain, nonatomic) EdgeDownloader *follows;
@property (retain, nonatomic) UISearchBar * searchBar;
@property (retain, nonatomic) EGFUser * selectedUser;
@end

@implementation UsersController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _searchBar = [[UISearchBar alloc] init];
    _searchBar.delegate = self;
    _searchBar.showsCancelButton = true;
    _searchBar.tintColor = [UIColor hexColor:0x5E66B1];
    
    EGF2SearchParameters * parameters = [EGF2SearchParameters parametersWithObject:@"user"];
    parameters.fields = @[@"first_name", @"last_name"];
    _searching = [[SearchDownloader alloc] initWithParameters:parameters];
    
    [self.tableView registerNib:[UINib nibWithNibName:@"ProgressCell" bundle:nil] forCellReuseIdentifier:@"ProgressCell"];
    [self.tableView registerNib:[UINib nibWithNibName:@"UserCell" bundle:nil] forCellReuseIdentifier:@"UserCell"];

    [self.graph userObjectWithCompletion:^(NSObject * object, NSError * error) {
        if ([object isKindOfClass:[EGFUser class]]) {
            EGFUser * user = (EGFUser *)object;
            _follows = [[EdgeDownloader alloc] initWithSource:user.id edge:@"follows" expand:@[]];
            _follows.tableView = self.tableView;
            [_follows getNextPage];
            _activeDownloader = _follows;
        }
    }];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if (_selectedUser) {
        NSInteger index = [_activeDownloader.graphObjects indexOfObject:_selectedUser];
        
        if (index != NSNotFound) {
            [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForItem:index inSection:0]] withRowAnimation:UITableViewRowAnimationNone];
        }
        _selectedUser = nil;
    }
}

- (void)setActiveDownloader:(EdgeDownloader *)activeDownloader {
    if (_activeDownloader) {
        _activeDownloader.tableView = nil;
    }
    _activeDownloader = activeDownloader;
    
    if (_activeDownloader) {
        _activeDownloader.tableView = self.tableView;
    }
}

- (IBAction)beginSearch:(id)sender {
    self.activeDownloader = _searching;
    self.navigationItem.rightBarButtonItem = nil;
    self.navigationItem.titleView = _searchBar;
    [_searchBar becomeFirstResponder];
}

- (void)endSearch {
    self.activeDownloader = _follows;
    _searchBar.text = nil;
    self.navigationItem.rightBarButtonItem = _searchButton;
    self.navigationItem.titleView = nil;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.destinationViewController isMemberOfClass:[UserProfileController class]] && [sender isMemberOfClass:[NSIndexPath class]]) {
        UserProfileController * controller = (UserProfileController *)segue.destinationViewController;
        NSIndexPath * indexPath = (NSIndexPath *)sender;
        _selectedUser = (EGFUser *)[_activeDownloader objectAtIndex:indexPath.row];
        controller.profileUser = _selectedUser;
        [_searchBar resignFirstResponder];
    }
}

// MARK:- UserCellDelegate
- (void)didTapFollowButtonWithUser:(EGFUser *)user andCell:(UserCell *)cell {
    if ([[Follows shared] followStateForUser:user] == isFollow) {
        [[Follows shared] unfollowUser:user completion:^{
            cell.user = user;
        }];
    }
    else {
        [[Follows shared] followUser:user completion:^{
            cell.user = user;
        }];
    }
    cell.user = user;
}

// MARK:- UISearchBarDelegate
-(void)searchBarCancelButtonClicked:(UISearchBar *)searchBar {
    [self endSearch];
}

- (void)searchBarTextDidEndEditing:(UISearchBar *)searchBar {
    if (searchBar.text.length == 0) {
        [self endSearch];
    }
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
    [_searching showObjectsWithQuery:searchText];
}
                
// MARK:- UIScrollViewDelegate
- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView {
    if (self.refreshControl.isRefreshing) {
        [_activeDownloader refreshList];
    }
}

// MARK:- UITableViewDelegate
- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 1) {
        if (![_activeDownloader isDownloaded]) {
            [_activeDownloader getNextPage];
        }
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [self performSegueWithIdentifier:@"ShowUserProfile" sender:indexPath];
}

// MARK:- UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (section == 0) {
        return _activeDownloader.count;
    }
    return 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 1) {
        ProgressCell * cell = [tableView dequeueReusableCellWithIdentifier:@"ProgressCell"];
        cell.indicatorIsHidden = tableView.refreshControl.isRefreshing || [_activeDownloader isDownloaded];
        return cell;
    }
    UserCell * cell = [tableView dequeueReusableCellWithIdentifier:@"UserCell"];
    cell.user = (EGFUser *)[_activeDownloader objectAtIndex:indexPath.row];
    cell.delegate = self;
    return cell;
}
@end

File: GuideSwift/GuideSwift/ViewControllers/UsersController/UsersController.swift
//
//  UsersController.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 20.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit
import EGF2

class UsersController: BaseTableController, UserCellDelegate, UISearchBarDelegate {
    
    @IBOutlet var searchButton: UIBarButtonItem!
    
    fileprivate var searching: SearchDownloader<EGFUser>?
    fileprivate var follows: EdgeDownloader<EGFUser>?
    fileprivate let searchBar = UISearchBar()
    fileprivate var selectedUser: EGFUser?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        tableView.register(UINib(nibName: "ProgressCell", bundle: nil), forCellReuseIdentifier: "ProgressCell")
        tableView.register(UINib(nibName: "UserCell", bundle: nil), forCellReuseIdentifier: "UserCell")
        
        let parameters = EGF2SearchParameters(withObject: "user")
        parameters.fields = ["first_name","last_name"]
        searching = SearchDownloader(withParameters: parameters)

        searchBar.delegate = self
        searchBar.showsCancelButton = true
        searchBar.tintColor = UIColor.hexColor(0x5E66B1)
        
        Graph.userObject { (object, error) in
            guard let user = object as? EGFUser, let userId = user.id else { return }
            self.follows = EdgeDownloader(withSource: userId, edge: "follows", expand: [])
            self.follows?.tableView = self.tableView
            self.follows?.getNextPage()
            self.activeDownloader = self.follows
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if let user = selectedUser, let index = activeDownloader?.graphObjects.index(of: user) {
            tableView.reloadRows(at: [IndexPath(row: index, section: 0)], with: .none)
            selectedUser = nil
        }
    }
    
    fileprivate var activeDownloader: BaseDownloader<EGFUser>? {
        willSet {
            if let oldDownloader = activeDownloader {
                oldDownloader.tableView = nil
            }
        }
        didSet {
            activeDownloader?.tableView = tableView
        }
    }
    
    @IBAction func beginSearch(_ sender: AnyObject) {
        activeDownloader = searching
        navigationItem.rightBarButtonItem = nil
        navigationItem.titleView = searchBar
        searchBar.becomeFirstResponder()
    }
    
    func endSearch() {
        activeDownloader = follows
        searchBar.text = nil
        navigationItem.rightBarButtonItem = searchButton
        navigationItem.titleView = nil
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let useerProfileController = segue.destination as? UserProfileController, let indexPath = sender as? IndexPath {
            selectedUser = activeDownloader?[indexPath.row]
            useerProfileController.profileUser = selectedUser
            searchBar.resignFirstResponder()
        }
    }
    
    // MARK:- UserCellDelegate
    func didTapFollowButton(withUser user: EGFUser, andCell cell: UserCell) {
        if Follows.shared.followState(forUser: user) == .isFollow {
            Follows.shared.unfollow(user: user) { cell.user = user }
        }
        else {
            Follows.shared.follow(user: user) { cell.user = user }
        }
        cell.user = user
    }
    
    // MARK:- UISearchBarDelegate
    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        endSearch()
    }
    
    func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
        if let text = searchBar.text, text.isEmpty {
            endSearch()
        }
    }
    
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        searching?.showObjects(withQuery: searchText)
    }
    
    // MARK:- UITableViewDelegate
    override func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        guard let control = refreshControl, control.isRefreshing == true else { return }
        activeDownloader?.refreshList()
    }
    
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        if indexPath.section == 1 {
            if let downloader = activeDownloader, !downloader.isDownloaded {
                downloader.getNextPage()
            }
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        performSegue(withIdentifier: "ShowUserProfile", sender: indexPath)
    }
    
    // MARK:- UITableViewDataSource
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 2
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == 0 {
            return activeDownloader?.count ?? 0
        }
        return 1
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 1 {
            let cell = tableView.dequeueReusableCell(withIdentifier: "ProgressCell") as! ProgressCell
            cell.indicatorIsHidden = tableView.refreshControl!.isRefreshing || (activeDownloader?.isDownloaded ?? true)
            return cell
        }
        let cell = tableView.dequeueReusableCell(withIdentifier: "UserCell") as! UserCell
        cell.user = activeDownloader![indexPath.row]
        cell.delegate = self
        return cell
    }
}
